src/python/pants/backend/python/util_rules/entry_points.py

```diff
@@ -0,0 +1,388 @@
+        Get(
+            ExplicitlyProvidedDependencies,
+            DependenciesRequest(tgt[PythonDistributionDependenciesField]),
+        )
+        for tgt in request.targets
+    )
+    resolved_entry_points = await MultiGet(
+        Get(
+            ResolvedPythonDistributionEntryPoints,
+            ResolvePythonDistributionEntryPointsRequest(tgt[PythonDistributionEntryPointsField]),
+        )
+        for tgt in request.targets
+    )
+
+    filtered_entry_point_pex_addresses: list[Address] = []
+    filtered_entry_point_modules: list[
+        tuple[Address, str, str, EntryPoint, ExplicitlyProvidedDependencies]
+    ] = []
+
+    distribution_entry_points: ResolvedPythonDistributionEntryPoints
+    for tgt, distribution_entry_points, explicitly_provided_deps in zip(
+        request.targets, resolved_entry_points, all_explicit_dependencies
+    ):
+        # use .val instead of .explicit_modules and .pex_binary_addresses to facilitate filtering
+        for ep_group, entry_points in distribution_entry_points.val.items():
+            if not request.group_predicate(tgt, ep_group):
+                continue
+            for ep_name, ep_val in entry_points.items():
+                if not request.predicate(tgt, ep_group, ep_name):
+                    continue
+                if ep_val.pex_binary_address:
+                    filtered_entry_point_pex_addresses.append(ep_val.pex_binary_address)
+                else:
+                    filtered_entry_point_modules.append(
+                        (
+                            tgt.address,
+                            ep_group,
+                            ep_name,
+                            ep_val.entry_point,
+                            explicitly_provided_deps,
+                        )
+                    )
+    filtered_module_owners = await MultiGet(
+        Get(PythonModuleOwners, PythonModuleOwnersRequest(entry_point.module, resolve=None))
+        for _, _, _, entry_point, _ in filtered_entry_point_modules
+    )
+
+    filtered_unambiguous_module_owners: OrderedSet[Address] = OrderedSet()
+    for (address, ep_group, ep_name, entry_point, explicitly_provided_deps), owners in zip(
+        filtered_entry_point_modules, filtered_module_owners
+    ):
+        filtered_unambiguous_module_owners.update(
+            get_python_distribution_entry_point_unambiguous_module_owners(
+                address, ep_group, ep_name, entry_point, explicitly_provided_deps, owners
+            )
+        )
+
+    return EntryPointDependencies(
+        Addresses((*filtered_entry_point_pex_addresses, *filtered_unambiguous_module_owners))
+    )
+
+
+async def _get_entry_point_deps_targets_and_predicates(
+    owning_address: Address, entry_point_deps: PythonTestsEntryPointDependenciesField
+) -> tuple[
+    Targets, PythonDistributionEntryPointGroupPredicate, PythonDistributionEntryPointPredicate
+]:
+    assert entry_point_deps.value, "Unexpected empty entry_point_dependencies field"
```

Comment:
How does this manifest to the user? (i.e. is this possible with BUILD file (mis)configuration? or is it merely a sign of an internal pants bug?)

What I'd prefer is to have the error message convey if this is in any way fixable when encountered, or if it's an internal bug and they need to turn to the pants project for assistance/report a bug.

Reply:
This serves to narrow the types for mypy. Everywhere that could call this rule bails out before calling it if the field is empty (See `infer_entry_point_dependencies` and `generate_entry_points_txt_from_entry_point_dependencies`). So, I expect this to only ever happen during development, not a released version.

Maybe I could raise an exception like `InvalidField`. Would that be better?

Reply:
I turned this into a more verbose `ValueError`. Thanks for the suggestion.

Reply:
Ah, right. In that case, I think an assert is just fine, as we don't expect to ever hit it, and it's presence is only to satisfy mypy. I've done the same, but leaving a hint as to why it's there (and thus, no assert message, as it's not supposed to ever hit): https://github.com/pantsbuild/pants/blob/e6b377da6ebfc06d251fc15b499c01f9c959efa4/src/python/pants/engine/internals/parametrize.py#L235